README.md
# AdjeminPay SDK PHP

[![Latest Stable Version](http://poser.pugx.org/adjeminpay/adjeminpay-sdk-php/v)](https://packagist.org/packages/adjeminpay/adjeminpay-sdk-php) [![Total Downloads](http://poser.pugx.org/adjeminpay/adjeminpay-sdk-php/downloads)](https://packagist.org/packages/adjeminpay/adjeminpay-sdk-php) [![Latest Unstable Version](http://poser.pugx.org/adjeminpay/adjeminpay-sdk-php/v/unstable)](https://packagist.org/packages/adjeminpay/adjeminpay-sdk-php) [![License](http://poser.pugx.org/adjeminpay/adjeminpay-sdk-php/license)](https://packagist.org/packages/adjeminpay/adjeminpay-sdk-php) [![PHP Version Require](http://poser.pugx.org/adjeminpay/adjeminpay-sdk-php/require/php)](https://packagist.org/packages/adjeminpay/adjeminpay-sdk-php)

The AdjeminPay PHP SDK provides convenient access to the AdjeminPay API from
applications written in the PHP language. It includes a pre-defined set of
classes for API resources that initialize themselves dynamically from API
responses which makes it compatible with a wide range of versions of the AdjeminPay API

## Requirements

PHP 7.4.0 and later.

## Composer

You can install the bindings via [Composer](http://getcomposer.org/). Run the following command:

```bash
composer require adjeminpay/adjeminpay-sdk-php
```

## Create a checkout
```php
<?php

use AdjeminPay\AdjeminPay;
use AdjeminPay\Transaction;

//Create AdjeminPay instance
$clientId = "CLIENT_ID"; //Client ID of an application created on  Merchant backoffice
$clientSecret  = "CLIENT_SECRET"; //Client Secret of an application created on  Merchant backoffice
$adjeminPay = new AdjeminPay($clientId, $clientSecret);

//Create a checkout
/** @var Transaction $transaction Transaction*/
$transaction = $adjeminPay->createCheckout([
  'amount' => 200, //required
  'currency_code' => 'XOF', //required
  'merchant_trans_id' => 'b72e51dc-7211-4e85-a937-5372c8769d36', //required You provide a merchant_trans_id
  'designation' => 'Paiement de facture', //required
  'customer_recipient_number' => '2250505000000', //required
  "customer_email" =>"customer@gmail.com",
  "customer_firstname" =>"Paul",
  "customer_lastname" =>"Koffi",
  "webhook_url":"https://example.com/webhook_url",
  "return_url": "https://example.com/success",
  "cancel_url": "https://example.com/failure"
]);


//Complete the checkout
/** @var Transaction $transaction Transaction*/
$transaction = $adjeminPay->completeCheckout('b72e51dc-7211-4e85-a937-5372c8769d36',[
  'operator_code' => 'wave_ci', //Your get all payments Methods with $adjeminPay->getPaymentMethods('CI')
  'customer_recipient_number' => '2250505000000', //required
  "customer_email" =>"customer@gmail.com",
  "customer_firstname" =>"Paul",
  "customer_lastname" =>"Koffi",
]);

var_dump($transaction);

```

## Payment Status
```php
<?php

use AdjeminPay\AdjeminPay;
use AdjeminPay\Transaction;

//Create AdjeminPay instance
$clientId = "CLIENT_ID"; //Client ID of an application created on  Merchant backoffice
$clientSecret  = "CLIENT_SECRET"; //Client Secret of an application created on  Merchant backoffice
$adjeminPay = new AdjeminPay($clientId, $clientSecret);

//Get Checkout or Payout Status by merchant_transaction_id
/** @var Transaction $transaction Transaction*/
$merchant_transaction_id = 'b72e51dc-7211-4e85-a937-5372c8769d36';
$transaction = $adjeminPay->getPaymentStatus($merchant_transaction_id);

if($transaction->getStatus() == Transaction::SUCCESSFUL){
    echo  "Transaction is successful"
}

if($transaction->getStatus() == Transaction::PENDING){
    echo  "Transaction is pending"
}

if($transaction->getStatus() == Transaction::FAILED){
    echo  "Transaction is failed"
}

var_dump($transaction);

```

## Create a Payout
```php
<?php

use AdjeminPay\AdjeminPay;
use AdjeminPay\Transaction;

//Create AdjeminPay instance
$clientId = "CLIENT_ID"; //Client ID of an application created on  Merchant backoffice
$clientSecret  = "CLIENT_SECRET"; //Client Secret of an application created on  Merchant backoffice
$adjeminPay = new AdjeminPay($clientId, $clientSecret);

//Create a Payout
/** @var Transaction $transaction Transaction*/
$transaction = $adjeminPay->createPayout([
  'operator_code' => 'wave_ci',
  'amount' => 200000, //required
  'currency_code' => 'XOF', //required
  'merchant_trans_id' => 'b72e51dc-7211-4e85-a937-5372c8769d36', //required You provide a merchant_trans_id
  'designation' => 'Paiement de salaire', //required
  'customer_recipient_number' => '2250505000000', //required
  "customer_email" =>"customer@gmail.com",
  "customer_firstname" =>"Paul",
  "customer_lastname" =>"Koffi",
  "webhook_url":"https://example.com/webhook_url"
]);


```


## Get Payment Methods by Country ISO Code
```php
<?php

use AdjeminPay\AdjeminPay;

//Create AdjeminPay instance
$clientId = "CLIENT_ID"; //Client ID of an application created on  Merchant backoffice
$clientSecret  = "CLIENT_SECRET"; //Client Secret of an application created on  Merchant backoffice
$adjeminPay = new AdjeminPay($clientId, $clientSecret);


//Get Payment Methods by Country CODE
$country_iso = 'CI';
$paymentMethods = $adjeminPay->getPaymentMethods($country_iso);

var_dump($paymentMethods);

```


CHANGELOG.md
# ChangeLog AdjeminPay SDK PHP

## [1.0.2] - 01/12/2023
* Guzzle updated to ^7.2

## [1.0.1] - 01/12/2023
* Guzzle updated

## [1.0.0] - 01/12/2023

* Create a Checkout
* Complete the Checkout
* Create a Payout
* Get Payment Status

test.php
<?php

use AdjeminPay\AdjeminPay;
use AdjeminPay\Transaction;

//Create AdjeminPay instance
$clientId = "CLIENT_ID"; //Client ID of an application created on  Merchant backoffice
$clientSecret  = "CLIENT_SECRET"; //Client Secret of an application created on  Merchant backoffice
$adjeminPay = new AdjeminPay($clientId, $clientSecret);


//Get Checkout or Payout Status by merchant_transaction_id
/** @var Transaction $transaction Transaction*/
$merchant_transaction_id = 'b72e51dc-7211-4e85-a937-5372c8769d36';
try {
    $transaction = $adjeminPay->getPaymentStatus($merchant_transaction_id);
} catch (Exception $e) {
    echo  "Error: ".$e->getMessage();
}

var_dump($transaction);




tests/AdjeminPayTest.php
<?php

namespace tests;

use AdjeminPay\AdjeminPay;
use AdjeminPay\Transaction;
use PHPUnit\Framework\TestCase;

class AdjeminPayTest extends TestCase{

    protected AdjeminPay $adjeminPay;

    /**
     *
     * @throws \AdjeminPay\Exception\AdjeminPayAuthException
     */
    public function setUp(): void
    {
        //Create AdjeminPay instance
        $clientId = "CLIENT_ID"; //Client ID of an application created on  Merchant backoffice
        $clientSecret  = "CLIENT_SECRET"; //Client Secret of an application created on  Merchant backoffice
        $this->adjeminPay = new AdjeminPay($clientId, $clientSecret);
    }

    /**
     * @throws \AdjeminPay\Exception\AdjeminPayArgumentException
     * @throws \AdjeminPay\Exception\AdjeminPayException
     * @throws \AdjeminPay\Exception\AdjeminPayAuthException
     */
    public function  testGetPaymentStatus():void{

        //Get Checkout or Payout Status by merchant_transaction_id
        /** @var Transaction $transaction Transaction*/
        $merchant_transaction_id = '1b56c3fb-24f5-407e-8779-f239ed086e6e';
        $transaction = $this->adjeminPay->getPaymentStatus($merchant_transaction_id);
        $this->assertInstanceOf(Transaction::class, $transaction);
        var_dump($transaction);

    }

}

tests/AdjeminPay/AdjeminPayTest.php
<?php

namespace tests\AdjeminPay;

use AdjeminPay\AdjeminPay;
use AdjeminPay\Transaction;
use PHPUnit\Framework\TestCase;

class AdjeminPayTest extends TestCase{

    protected AdjeminPay $adjeminPay;

    /**
     *
     * @throws \AdjeminPay\Exception\AdjeminPayAuthException
     */
    public function setUp(): void
    {
        //Create AdjeminPay instance
        $clientId = "9a9b2458-5f5f-4e10-bd66-6ac41a4de7b2"; //Client ID of an application created on  Merchant backoffice
        //$clientId = "CLIENT_ID"; //Client ID of an application created on  Merchant backoffice
        $clientSecret  = "MlXk0cGHhHXQdXgEXoyy9qOGxQPWNJRjCIxufFoZ"; //Client Secret of an application created on  Merchant backoffice
        //$clientSecret  = "CLIENT_SECRET"; //Client Secret of an application created on  Merchant backoffice
        $this->adjeminPay = new AdjeminPay($clientId, $clientSecret);
    }

    /**
     * @throws \AdjeminPay\Exception\AdjeminPayArgumentException
     * @throws \AdjeminPay\Exception\AdjeminPayException
     * @throws \AdjeminPay\Exception\AdjeminPayAuthException
     */
    public function  testGetPaymentStatus():void{

        //Get Checkout or Payout Status by merchant_transaction_id
        /** @var Transaction $transaction Transaction*/
        $merchant_transaction_id = '1b56c3fb-24f5-407e-8779-f239ed086e6e';
        $transaction = $this->adjeminPay->getPaymentStatus($merchant_transaction_id);
        $this->assertInstanceOf(Transaction::class, $transaction);
        var_dump($transaction);

    }

}

src/AdjeminPay/StatusCode.php
<?php

namespace AdjeminPay;


class StatusCode
{

    const  OK = "OK";
    const  SUCCESS = "SUCCESS";
    const  OPERATION_ERROR = "OPERATION_ERROR";
    const  NOT_FOUND = "NOT_FOUND";
    const  INVALID_CREDENTIALS  = "INVALID_CREDENTIALS";
    const  INVALID_PARAMS = "INVALID_PARAMS";
    const  EXPIRED_TOKEN = "EXPIRED_TOKEN";
    const  INVALID_TOKEN = "INVALID_TOKEN";
    const  TRANSACTION_EXIST = "TRANSACTION_EXIST";
    const  INITIATED = "INITIATED";
    const  PENDING = "PENDING";
    const  EXPIRED = "EXPIRED";
    const  OTP_ERROR = "OTP_ERROR";
    const  OTP_EXPIRED = "OTP_EXPIRED";
    const  INSUFFICIENT_BALANCE = "INSUFFICIENT_BALANCE";
    const  USER_NOT_FOUND = "USER_NOT_FOUND";
    const  USER_IS_BLOCKED = "USER_IS_BLOCKED";
    const  FAILED = "FAILED";
    const  NOT_ALLOWED= "NOT_ALLOWED";
    const  CANCELLED = "CANCELLED";
    const  BAD_REQUEST  = "BAD_REQUEST";
    const  UNIMPLEMENTED  = "UNIMPLEMENTED";
    const  UNAUTHORIZED = "UNAUTHORIZED";


    const  messages  = [
        'OK'=>"Successful operation",
        'SUCCESS'=>"Transaction is successfully processed",
        "OPERATION_ERROR"=>"An error has occurred",
        "NOT_FOUND"=>"Not found",
        'BAD_REQUEST' => 'Your request is missing some headers or parameters',
        'INVALID_CREDENTIALS'=>'The requested service needs credentials, but the ones provided were invalid.',
        'INVALID_PARAMS'=>"Params you provides are invalid",
        'EXPIRED_TOKEN'=>"Token has expired",
        'INVALID_TOKEN'=>"Token is invalid",
        'TRANSACTION_EXIST'=>'The transaction already exists',
        'INITIATED'=>"Waiting for user entry",
        'PENDING'=>'User have started payment',
        'EXPIRED'=>'User has not confirmed the payment',
        'OTP_ERROR'=>'Otp user provided is incorrect',
        'OTP_EXPIRED'=>'Otp user provided has expired',
        'INSUFFICIENT_BALANCE'=>'User has not enough balance to validate operation',
        'USER_NOT_FOUND'=>'User does not exist',
        'USER_IS_BLOCKED'=>'User has been blocked',
        'FAILED'=>'Payment has failed',
        'NOT_ALLOWED'=>'This Ip is not whitelisted',
        'CANCELLED' => 'User have cancelled payment',
        'UNIMPLEMENTED' => 'Unimplemented',
        'UNAUTHORIZED' => 'Authorization Token not found'
    ];

    const  codes  = [
        'OK'=>200,
        'SUCCESS'=>100,
        "OPERATION_ERROR"=>500,
        "NOT_FOUND"=>404,
        'UNAUTHORIZED'=>401,
        'BAD_REQUEST'=>400,
        'INVALID_CREDENTIALS'=>1005,
        'INVALID_PARAMS'=>400,
        'EXPIRED_TOKEN'=>1003,
        'INVALID_TOKEN'=>1004,
        'TRANSACTION_EXIST'=>1200,
        'INITIATED'=>2001,
        'PENDING'=>2002,
        'EXPIRED'=>2003,
        'OTP_ERROR'=>2004,
        'OTP_EXPIRED'=>2008,
        'INSUFFICIENT_BALANCE'=>2005,
        'USER_NOT_FOUND'=>2006,
        'USER_IS_BLOCKED'=>2007,
        'FAILED'=>2010,
        'NOT_ALLOWED'=>2011,
        'CANCELLED' => 2012,
        'UNIMPLEMENTED' => 2013
    ];

    const  statuses  = [
        200 => 'OK',
        100 => 'SUCCESS',
        1 => "OPERATION_ERROR",
        404 => "NOT_FOUND",
        400 => 'BAD_REQUEST',
        1005 => 'INVALID_CREDENTIALS',
        1002 => 'INVALID_PARAMS',
        1003 => 'EXPIRED_TOKEN',
        1004 => 'INVALID_TOKEN',
        1200 => 'TRANSACTION_EXIST',
        2001 => 'INITIATED',
        2002 => 'PENDING',
        2003 => 'EXPIRED',
        2004 => 'OTP_ERROR',
        2008 => 'OTP_EXPIRED',
        2005 => 'INSUFFICIENT_BALANCE',
        2006 => 'USER_NOT_FOUND',
        2007 => 'USER_IS_BLOCKED' ,
        2010 => 'FAILED',
        2011 => 'NOT_ALLOWED',
        2012 => 'CANCELLED',
        2013 => 'UNIMPLEMENTED'
    ];

}


src/AdjeminPay/AdjeminPay.php
<?php

namespace AdjeminPay;

use AdjeminPay\Exception\AdjeminPayArgumentException;
use AdjeminPay\Exception\AdjeminPayAuthException;
use GuzzleHttp\Client;
use AdjeminPay\Exception\AdjeminPayException;

/**
 * AdjeminPay Class
 *
 * @version 1.0.0
 */
class AdjeminPay implements AdjeminPayInterface {

    const API_BASE_URL = "https://api.adjeminpay.com";

    /**
     * @var string $clientId Client ID
     */
    private $clientId;

    /**
     * @var string $clientSecret Client Secret
     */
    private $clientSecret;

    /**
     * @var array $data All information about the application or transaction
     */
    public $data;

    /**
     * @var string $token Access token
     */
    private $token;

    /**
     * @var array $response Transaction reponse data
     */
    private $response;

    /**
     * Class constructor
     * Initialize some private value and check if they are available
     *
     * @param string $clientId
     * @param string $clientSecret
     * @throws AdjeminPayAuthException
     */
    public function __construct($clientId, $clientSecret){
        $this->clientId = $clientId;
        $this->clientSecret = $clientSecret;
        $this->token = $this->obtainAccessToken();
    }

    /**
     * @return string
     */
    public function getBaseUrl()
    {
        return self::API_BASE_URL;
    }

    /**
     * @return string
     */
    public function getClientId()
    {
        return $this->clientId;
    }

    /**
     * @return string
     */
    public function getClientSecret()
    {
        return $this->clientSecret;
    }

    public function obtainAccessToken()
    {
        $client = new Client();
        $url = $this->getBaseUrl()."/oauth/token";
        $body = [
            'client_id' => ''.$this->getClientId(),
            'client_secret'=> ''.$this->getClientSecret(),
            'grant_type' => 'client_credentials'
        ];

        $response = $client->post($url, [
            "headers" => [
                'Accept' => 'application/json',
                'Content-Type' => 'application/x-www-form-urlencoded'
            ],
            "form_params" => $body

        ]);

        if ($response->getStatusCode() == 200){
            $body = $response->getBody()->getContents();
            $json = (array) json_decode($body, true);

            if(array_key_exists('access_token', $json) && !empty( $json['access_token'])){
                return $json['access_token'];
            }else{
                if(array_key_exists('message', $json) && !empty( $json['message'])){
                    $message  = $json['message'];
                }else{
                    $message  = "Client authentication failed";
                }
                throw  new AdjeminPayAuthException($message,$response->getStatusCode());
            }

        }else{
            $body = $response->getBody()->getContents();
            $json = (array) json_decode($body, true);
            if(array_key_exists('message', $json) && !empty( $json['message'])){
                $message  = $json['message'];
            }else{
                $message  = "Client authentication failed";
            }
            throw  new AdjeminPayAuthException($message,$response->getStatusCode());
        }
    }

    /**
     * Obtain Access Token
     * @throws AdjeminPayAuthException
     */
    public function getAccessToken()
    {
        return $this->token;
    }

    /**
     * @throws AdjeminPayException
     * @throws AdjeminPayAuthException
     */
    public function createCheckout($params): Transaction
    {
        //Validate parameters
        if(!array_key_exists('amount',$params) || empty($params['amount'])) throw new AdjeminPayArgumentException("Bad request,  amount is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);
        if(!array_key_exists('currency_code',$params) || empty($params['currency_code'])) throw new AdjeminPayArgumentException("Bad request,  currency_code is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);
        if(!array_key_exists('designation',$params) || empty($params['designation'])) throw new AdjeminPayArgumentException("Bad request,  designation is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);
        if(!array_key_exists('merchant_trans_id',$params) || empty($params['merchant_trans_id'])) throw new AdjeminPayArgumentException("Bad request,  merchant_trans_id is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);

        if(empty($this->getAccessToken())){
            $message = 'The requested service needs credentials, but the ones provided were invalid.';
            throw  new AdjeminPayAuthException($message,401);
        }

        $client = new Client();
        $url = $this->getBaseUrl()."/v3/merchants/create_checkout";
        $body = [
            'amount' => intval($params['amount']),
            'currency_code' => array_key_exists('currency_code', $params)?$params['currency_code']:null,
            'merchant_trans_id' => array_key_exists('merchant_trans_id', $params)?$params['merchant_trans_id']:null,
            'merchant_trans_data' => array_key_exists('merchant_trans_data', $params)?$params['merchant_trans_data']:null,
            'designation' => array_key_exists('designation', $params)?$params['designation']:null,
            'webhook_url' => array_key_exists('webhook_url', $params)?$params['webhook_url']:null,
            'return_url' => array_key_exists('return_url', $params)?$params['return_url']:null,
            'cancel_url' => array_key_exists('cancel_url', $params)?$params['cancel_url']:null,
            'customer_recipient_number' => array_key_exists('customer_recipient_number', $params)?$params['customer_recipient_number']:null,
            'customer_email' => array_key_exists('customer_email', $params)?$params['customer_email']:null,
            'customer_firstname' => array_key_exists('customer_firstname', $params)?$params['customer_firstname']:null,
            'customer_lastname' => array_key_exists('customer_lastname', $params)?$params['customer_lastname']:null,
        ];

        $response = $client->post($url, [
            "headers" => [
                'Authorization' => 'Bearer '.$this->getAccessToken(),
                'Accept' => 'application/json',
                'Content-Type' => 'application/json'
            ],
            "json" => $body

        ]);

        $body = $response->getBody()->getContents();
        if($response->getStatusCode() == 200){
            $json = json_decode($body, true);

            if(array_key_exists('data', $json) && !empty( $json['data'])){
                $data = $json['data'];
                return new TransactionImpl($data);
            }else{
                $json =  json_decode($body, true);
                if(array_key_exists('message', $json) && !empty( $json['message'])){
                    $message  = $json['message'];
                }else{
                    $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
                }
                throw  new AdjeminPayException($message,$response->getStatusCode());
            }

        }else{

            $json =  json_decode($body, true);
            if(array_key_exists('message', $json) && !empty( $json['message'])){
                $message  = $json['message'];
            }else{
                $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
            }
            throw  new AdjeminPayException($message,$response->getStatusCode());
        }
    }

    public function completeCheckout($merchant_trans_id, $params): Transaction
    {

        //Validate parameters
        if(!array_key_exists('operator_code',$params) || empty($params['operator_code'])) throw new AdjeminPayArgumentException("Bad request,  operator_code is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);
        if(!array_key_exists('customer_recipient_number',$params) || empty($params['customer_recipient_number'])) throw new AdjeminPayArgumentException("Bad request,  customer_recipient_number is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);
        if( empty($merchant_trans_id)) throw new AdjeminPayArgumentException("Bad request,  merchant_trans_id is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);

        if(empty($this->getAccessToken())){
            $message = 'The requested service needs credentials, but the ones provided were invalid.';
            throw  new AdjeminPayAuthException($message,401);
        }

        $client = new Client();
        $url = $this->getBaseUrl()."/v3/merchants/complete_checkout/".$merchant_trans_id;
        $body = [
            'operator_code' => intval($params['operator_code']),
            'customer_recipient_number' => array_key_exists('customer_recipient_number', $params)?$params['customer_recipient_number']:null,
            'customer_email' => array_key_exists('customer_email', $params)?$params['customer_email']:null,
            'customer_firstname' => array_key_exists('customer_firstname', $params)?$params['customer_firstname']:null,
            'customer_lastname' => array_key_exists('customer_lastname', $params)?$params['customer_lastname']:null,
            'otp' => array_key_exists('otp', $params)?$params['otp']:null
        ];

        $response = $client->post($url, [
            "headers" => [
                'Authorization' => 'Bearer '.$this->getAccessToken(),
                'Accept' => 'application/json',
                'Content-Type' => 'application/json'
            ],
            "json" => $body

        ]);

        $body = $response->getBody()->getContents();
        if($response->getStatusCode() == 200){
            $json = json_decode($body, true);

            if(array_key_exists('data', $json) && !empty( $json['data'])){
                $data = $json['data'];
                return new TransactionImpl($data);
            }else{
                $json =  json_decode($body, true);
                if(array_key_exists('message', $json) && !empty( $json['message'])){
                    $message  = $json['message'];
                }else{
                    $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
                }
                throw  new AdjeminPayException($message,$response->getStatusCode());
            }

        }else{

            $json =  json_decode($body, true);
            if(array_key_exists('message', $json) && !empty( $json['message'])){
                $message  = $json['message'];
            }else{
                $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
            }
            throw  new AdjeminPayException($message,$response->getStatusCode());
        }
    }

    public function createPayout($params): Transaction
    {

        //Validate parameters
        if(!array_key_exists('amount',$params) || empty($params['amount'])) throw new AdjeminPayArgumentException("Bad request,  amount is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);
        if(!array_key_exists('currency_code',$params) || empty($params['currency_code'])) throw new AdjeminPayArgumentException("Bad request,  currency_code is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);
        if(!array_key_exists('designation',$params) || empty($params['designation'])) throw new AdjeminPayArgumentException("Bad request,  designation is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);
        if(!array_key_exists('merchant_trans_id',$params) || empty($params['merchant_trans_id'])) throw new AdjeminPayArgumentException("Bad request,  merchant_trans_id is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);
        if(!array_key_exists('customer_recipient_number',$params) || empty($params['customer_recipient_number'])) throw new AdjeminPayArgumentException("Bad request, customer_recipient_number is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);

        if(empty($this->getAccessToken())){
            $message = 'The requested service needs credentials, but the ones provided were invalid.';
            throw  new AdjeminPayAuthException($message,401);
        }

        $client = new Client();
        $url = $this->getBaseUrl()."/v3/merchants/create_payout";
        $body = [
            'amount' => intval($params['amount']),
            'currency_code' => array_key_exists('currency_code', $params)?$params['currency_code']:null,
            'merchant_trans_id' => array_key_exists('merchant_trans_id', $params)?$params['merchant_trans_id']:null,
            'merchant_trans_data' => array_key_exists('merchant_trans_data', $params)?$params['merchant_trans_data']:null,
            'designation' => array_key_exists('designation', $params)?$params['designation']:null,
            'webhook_url' => array_key_exists('webhook_url', $params)?$params['webhook_url']:null,
            'return_url' => array_key_exists('return_url', $params)?$params['return_url']:null,
            'cancel_url' => array_key_exists('cancel_url', $params)?$params['cancel_url']:null,
            'customer_recipient_number' => array_key_exists('customer_recipient_number', $params)?$params['customer_recipient_number']:null,
            'customer_email' => array_key_exists('customer_email', $params)?$params['customer_email']:null,
            'customer_firstname' => array_key_exists('customer_firstname', $params)?$params['customer_firstname']:null,
            'customer_lastname' => array_key_exists('customer_lastname', $params)?$params['customer_lastname']:null,
        ];

        $response = $client->post($url, [
            "headers" => [
                'Authorization' => 'Bearer '.$this->getAccessToken(),
                'Accept' => 'application/json',
                'Content-Type' => 'application/json'
            ],
            "json" => $body

        ]);

        $body = $response->getBody()->getContents();
        if($response->getStatusCode() == 200){
            $json = json_decode($body, true);

            if(array_key_exists('data', $json) && !empty( $json['data'])){
                $data = $json['data'];
                return new TransactionImpl($data);
            }else{
                $json =  json_decode($body, true);
                if(array_key_exists('message', $json) && !empty( $json['message'])){
                    $message  = $json['message'];
                }else{
                    $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
                }
                throw  new AdjeminPayException($message,$response->getStatusCode());
            }

        }else{

            $json =  json_decode($body, true);
            if(array_key_exists('message', $json) && !empty( $json['message'])){
                $message  = $json['message'];
            }else{
                $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
            }
            throw  new AdjeminPayException($message,$response->getStatusCode());
        }
    }

    public function getPaymentStatus($merchant_trans_id)
    {

        //Validate parameters
        if(empty($merchant_trans_id)) throw new AdjeminPayArgumentException("Bad request,  merchant_trans_id is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);

        if(empty($this->getAccessToken())){
            $message = 'The requested service needs credentials, but the ones provided were invalid.';
            throw  new AdjeminPayAuthException($message,401);
        }

        $client = new Client();
        $url = $this->getBaseUrl()."/v3/merchants/payment/$merchant_trans_id";

        $response = $client->get($url, [
            "headers" => [
                'Authorization' => 'Bearer '.$this->getAccessToken(),
                'Accept' => 'application/json'
            ]

        ]);

        $body = $response->getBody()->getContents();
        if($response->getStatusCode() == 200){
            $json = json_decode($body, true);

            if(array_key_exists('data', $json) && !empty( $json['data'])){
                $data = $json['data'];
                return new TransactionImpl($data);
            }else{
                $json =  json_decode($body, true);
                if(array_key_exists('message', $json) && !empty( $json['message'])){
                    $message  = $json['message'];
                }else{
                    $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
                }
                throw  new AdjeminPayException($message,$response->getStatusCode());
            }

        }else{

            $json =  json_decode($body, true);
            if(array_key_exists('message', $json) && !empty( $json['message'])){
                $message  = $json['message'];
            }else{
                $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
            }
            throw  new AdjeminPayException($message,$response->getStatusCode());
        }
    }

    public function getPaymentMethods($country_iso)
    {
        //Validate parameters
        if(empty($country_iso)) throw new AdjeminPayArgumentException("Bad request,  country_iso is required", StatusCode::codes[StatusCode::INVALID_PARAMS]);

        if(empty($this->getAccessToken())){
            $message = 'The requested service needs credentials, but the ones provided were invalid.';
            throw  new AdjeminPayAuthException($message,401);
        }

        $client = new Client();
        $url = $this->getBaseUrl()."/v3/operators/".strtoupper($country_iso);

        $response = $client->get($url, [
            "headers" => [
                'Authorization' => 'Bearer '.$this->getAccessToken(),
                'Accept' => 'application/json'
            ]
        ]);

        $body = $response->getBody()->getContents();
        if($response->getStatusCode() == 200){
            $json = json_decode($body, true);

            if(array_key_exists('data', $json) && !empty( $json['data'])){
                $data = $json['data'];
                if(!is_array($data)){
                     $data = (array)$data;
                }

                return $data;
            }else{
                $json =  json_decode($body, true);
                if(array_key_exists('message', $json) && !empty( $json['message'])){
                    $message  = $json['message'];
                }else{
                    $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
                }
                throw  new AdjeminPayException($message,$response->getStatusCode());
            }

        }else{

            $json =  json_decode($body, true);
            if(array_key_exists('message', $json) && !empty( $json['message'])){
                $message  = $json['message'];
            }else{
                $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
            }
            throw  new AdjeminPayException($message,$response->getStatusCode());
        }
    }

    public function getBalance()
    {

        if(empty($this->getAccessToken())){
            $message = 'The requested service needs credentials, but the ones provided were invalid.';
            throw  new AdjeminPayAuthException($message,401);
        }

        $client = new Client();
        $url = $this->getBaseUrl()."/v3/merchants/show";

        $response = $client->get($url, [
            "headers" => [
                'Authorization' => 'Bearer '.$this->getAccessToken(),
                'Accept' => 'application/json'
            ]
        ]);

        $body = $response->getBody()->getContents();
        if($response->getStatusCode() == 200){
            $json = json_decode($body, true);

            if(array_key_exists('data', $json) && !empty( $json['data'])){
                $data = $json['data'];
                if(!is_array($data)){
                    $data = (array)$data;
                }

                return $data;
            }else{
                $json =  json_decode($body, true);
                if(array_key_exists('message', $json) && !empty( $json['message'])){
                    $message  = $json['message'];
                }else{
                    $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
                }
                throw  new AdjeminPayException($message,$response->getStatusCode());
            }

        }else{

            $json =  json_decode($body, true);
            if(array_key_exists('message', $json) && !empty( $json['message'])){
                $message  = $json['message'];
            }else{
                $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
            }
            throw  new AdjeminPayException($message,$response->getStatusCode());
        }
    }

    public function getKnowIfWeCanCollectMoney()
    {
        if(empty($this->getAccessToken())){
            $message = 'The requested service needs credentials, but the ones provided were invalid.';
            throw  new AdjeminPayAuthException($message,401);
        }

        $client = new Client();
        $url = $this->getBaseUrl()."/v3/merchants/know_if_can_collect_money";

        $response = $client->get($url, [
            "headers" => [
                'Authorization' => 'Bearer '.$this->getAccessToken(),
                'Accept' => 'application/json'
            ]
        ]);

        $body = $response->getBody()->getContents();
        if($response->getStatusCode() == 200){
            $json = json_decode($body, true);

            if(array_key_exists('data', $json) && !empty( $json['data'])){
                $data = $json['data'];
                if(!is_bool($data)){
                    $data = boolval($data);
                }

                return $data;
            }else{
                $json =  json_decode($body, true);
                if(array_key_exists('message', $json) && !empty( $json['message'])){
                    $message  = $json['message'];
                }else{
                    $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
                }
                throw  new AdjeminPayException($message,$response->getStatusCode());
            }

        }else{

            $json =  json_decode($body, true);
            if(array_key_exists('message', $json) && !empty( $json['message'])){
                $message  = $json['message'];
            }else{
                $message  = StatusCode::messages[StatusCode::OPERATION_ERROR];
            }
            throw  new AdjeminPayException($message,$response->getStatusCode());
        }
    }
}